packages/account/src/BaseSmartAccount.ts

```diff
@@ -0,0 +1,537 @@
+import { JsonRpcProvider, Provider } from '@ethersproject/providers'
+import { BigNumber, BigNumberish, Signer, BytesLike, ethers, Bytes } from 'ethers'
+import { IBaseSmartAccount } from './interfaces/IBaseSmartAccount'
+import { defaultAbiCoder, keccak256, arrayify, Result } from 'ethers/lib/utils'
+import { UserOperation, ChainId } from '@biconomy/core-types'
+import { calcPreVerificationGas, DefaultGasLimits } from './utils/Preverificaiton'
+import { NotPromise, packUserOp } from '@biconomy/common'
+import { IBundler, UserOpResponse } from '@biconomy/bundler'
+import { IPaymaster, PaymasterAndDataResponse } from '@biconomy/paymaster'
+import { EntryPoint_v100, Logger } from '@biconomy/common'
+import { BaseSmartAccountConfig, Overrides, TransactionDetailsForUserOp } from './utils/Types'
+import { GasOverheads } from './utils/Preverificaiton'
+import { EntryPoint, EntryPoint__factory } from '@account-abstraction/contracts'
+import { DEFAULT_ENTRYPOINT_ADDRESS } from './utils/Constants'
+import { RPC_PROVIDER_URLS } from '@biconomy/common'
+
+type UserOperationKey = keyof UserOperation
+
+export abstract class BaseSmartAccount implements IBaseSmartAccount {
+  private isDeployed = false
+  bundler?: IBundler // httpRpcClient
+  paymaster?: IPaymaster // paymasterAPI
+  overheads?: Partial<GasOverheads>
+  entryPointAddress!: string
+  accountAddress?: string
+  owner?: Signer // owner is not mandatory for some account implementations
+  index?: number
+  chainId?: ChainId
+  provider: JsonRpcProvider | Provider // Review
```

Comment:
can make it Provider.

Reply:
I had some doubts here about should we make provider mandatory? currently in live sdk chainId is mandatory. what would be easier for devs? 

if provider is given use the provider
if rpcUrl is given then make JsonRpcProvider using this custom rpcUrl
if nothing is given chainId is needed to get public rpcUrl from constants and make JsonRpcProvider

Reply:
I don't think we should make Provider mandatory or even get it from the dapp. Coz for example in case of multiple chain support, there won't be an easy way to pass provider object for each network. 

chainId or rpcUrl are enough for SmartAccount package to operate.